docs: accts, idp, and disclosure component (#3102)
interledger/rafiki: packages/documentation/src/content/docs/integration/requirements/idp.mdx

An identity provider (IdP) is a system or service that stores and manages user identity information, authentication, and consent. Examples of IdPs include OpenID Connect and Okta.

Open Payments requires any authorization server that issues interactive grants be integrated with an IdP. Interactive grants are used to gather consent, and are required for Open Payments <LinkOut href="https://openpayments.dev/introduction/op-concepts/#outgoing-payment">outgoing payments</LinkOut>. Interactive grants are optional for incoming payments and quotes. 

Responsibilities of your IdP include:

> - Providing an interface to gather consent

Reviewer: ```suggestion
- Providing an interface to gather end-user consent for a particular action
```
just giving some more specific detail

- Sending the interaction choice (approve or deny) to your authorization server
- Sending a request to your authorization server to finish the interaction
- Redirecting the user after the interaction is complete

> - Authenticating requests from clients to create payments on Rafiki's backend

Reviewer: > Authenticating requests from clients to create payments on Rafiki’s backend

This is more RS & AS responsibility

Open Payments requires that consent be collected via an interactive grant before an outgoing payment request is issued. A grant is interactive when explicit interaction by an entity (e.g., the software's end user) is required to approve or deny the grant. Tapping an *Approve* button to authorize a payment is an example of an explicit interaction.

See the <LinkOut href="https://openpayments.dev/introduction/grants/">Open Payments documentation</LinkOut> for more information on grant negotiation and authorization.

## Authorization servers

Authorization servers grant permission to clients to access the Open Payments APIs. This enables clients to create incoming payments, quotes, and outgoing payments against an account holder's account.

Rafiki's [auth service](/integration/services/auth-service) provides you with a reference implementation of an Open Payments authorization server. You can use the service as an alternative to developing your own in-house solution. Additionally, the server extends an [API](#interaction-endpoints) that provides interaction endpoints for your IdP.

### Environment variables

> The following `backend` variables must be configured on your authorization server.

Reviewer: ```suggestion
The following `auth` variables must be configured on your authorization server.
```

<div class="overflow-table wide-column">

| Variable | Helm value name | Default | Description |
| -------- | --------------- | ------- | ----------- |
| `IDENTITY_SERVER_URL` | `auth.identityServer.domain` | N/A | Your IdP server's URL where your authorization server will direct clients to so end-users can complete the interaction and authorize a grant. | Y |
| `IDENTITY_SERVER_SECRET` | `auth.identityServer.secret` | N/A | A shared secret between your authorization and IdP servers that your authorization server will use to secure its IdP-related endpoints.<br/>When your IdP sends requests to your authorization server, your IdP must provide the secret via an [`x-idp-secret`](#x-idp-secret-header) header. | Y |

> | `INCOMING_PAYMENT_INTERACTION` | `auth.interaction.incomingPayment` | `false` | Indicates whether incoming payments grant requests are interactive. | Y |

Reviewer: We also have `QUOTE_INTERACTION`

Author: The table is for backend variables, so I just want to make sure that `QUOTE_INTERACTION` is an Auth variable (our environment variables page has it listed under Auth > Optional).

| `INTERACTION_EXPIRY_SECONDS` | `auth.interactionExpirySeconds` | `600` | The time in seconds for which a user can interact with a grant request | Y |
| `INTERACTION_PORT` | `auth.port.interaction` | `3009` | The port number for the [interaction endpoints](#interaction-endpoints) | Y |

> | `INTROSPECTION_PORT` | `auth.port.introspection` | `3007` | The port number of your Open Payments authorization token introspection server | Y |

Reviewer: I think we don't need to include the `INTROSPECTION_PORT` here - since it's not IdP specific

| `LIST_ALL_ACCESS_INTERACTION` | N/A | `true` | Specifies whether grant requests including a `list-all` action should require interaction. In these requests, the client asks to list resources that they themselves did not create. | Y |

</div>

## Interaction endpoints

The authorization server provided by Rafiki's `auth` service extends an API for your IdP server to use after a pending grant request is created.

Each interaction with an endpoint is identified by an `id` and a `nonce`. Both are provided as query parameters when your authorization server redirects to your IdP's server.

> The endpoints are tied to the `IDENTITY_SERVER_URL` you defined when configuring your environment variables. For example, if your identity server URL is `https://idp.wallet.example.com`, then calling the `/interact/{id}/{nonce}` endpoint to start a user interaction session would look as follows:

Reviewer: The endpoints are tied to the auth server URL. `IDENTITY_SERVER_URL` is used by the AS to inform the Open Payments client where to redirect the user (`interact.redirect` in the initial grant response)

```
https://idp.wallet.example.com/interact/{id}/{nonce}
```

### Interaction endpoints

> The endpoints are called in the sequence listed below.

Reviewer: I think we should specify who is calling which endpoints, as the grant continuation request is called by the Open Payments client instead of the IDP

| <Badge text="POST" variant="success" size="medium"/> | `/interact/{id}/{nonce}` | [Continue grant](#continue-grant) |

</div>

We also provide an <LinkOut href="https://github.com/interledger/rafiki/blob/main/packages/auth/src/openapi/specs/id-provider.yaml">OpenAPI specification</LinkOut> that describes the endpoints. Note that the *Continue grant* endpoint is not included in the spec because it's part of the <LinkOut href="https://openpayments.dev/apis/auth-server/operations/post-continue/">Open Payments Auth Server API</LinkOut>.

#### Start user interaction session

Called by the client to establish an interactive session with your authorization server. Also redirects the client's browser session to your IdP consent screen.

#### Look up grant information

Called by your IdP server to retrieve a list of access rights, requested by the client, from your authorization server. The request is secured with an [`x-idp-secret`](#x-idp-secret-header) header. The access rights are presented to the client's end-user on the consent screen. The authorization server's response is served on your configured `INTERACTION_PORT`.

#### Accept or reject grant

> Your IdP server communicates the choice made by the end-user on the consent screen (accept/reject) to your authorization server. The request is secured with an [`x-idp-secret`](#x-idp-secret-header) header. Then, your IdP server redirects to the `finish` endpoint. The response is served on your configured `INTERACTION_PORT`.

Reviewer: Instead of redirecting, its more so the IdP calls the `finish` endpoint on Rafiki's auth service (which handles the interaction requests on `INTERACTION_PORT`)

#### Finish interaction

Called by your IdP server to end the interaction and redirect the end-user's browser session to the URI of the <LinkOut href="https://openpayments.dev/apis/auth-server/operations/post-request/">grant initialization request</LinkOut>.

> The `result` query parameter indicates the success or failure of the grant authorization. When successful, the SHA-256 hash of the interaction is sent in the response along with an `interact_ref` that identifies the interaction on the authorization server and the URI of the grant initialization request.

Reviewer: We can specify something about the hash having to be verified by the original client (ie its not an IDP concern)

</div>

#### Continue grant

The client requests a grant from your authorization server for an accepted interaction. Your authorization server responds with an <LinkOut href="https://openpayments.dev/apis/auth-server/operations/post-continue/">access token</LinkOut>.

## x-idp-secret header

The `x-idp-secret` header is used for requests to the following endpoints:

- `GET /grant/:id/:nonce`
- `POST /grant/:id/:nonce/accept`
- `POST /grant/:id/:nonce/reject`

> The header's purpose is to secure communications between your IdP and authorization server and its value should be a shared secret known to both entities. When your IdP server sends requests to your authorization server, your IdP must provide the secret via this header.

Reviewer: Nit: 

> "When your IdP server sends requests to your authorization server" 

The `x-idp-secret` is specific to Rafiki's AS, and doesn't apply generally (ie if the ASE wants to run their own AS, they can add security in a way they see fit)

An identity provider (IdP) is a system or service that stores and manages user identity information, authentication, and consent. Examples of IdPs include OpenID Connect and Okta.

Open Payments requires any authorization server that issues interactive grants be integrated with an IdP. Interactive grants are used to gather consent. More information about interactive grants are available [below](#interactive-grants).

Responsibilities of your IdP include:
- Providing an interface to gather end-user consent for a particular action
- Sending the interaction choice (approve or deny) to your authorization server
- Sending a request to your authorization server to finish the interaction
- Redirecting the user after the interaction is complete

:::note
We provide Ory Kratos, a cloud-based user management system, for the identity and user management of your Rafiki Admin users. Kratos is for internal use only and **cannot** be used as your IdP for Open Payments.
:::

## Interactive grants

In Open Payments, grants are used to indicate a resource owner, such as an account holder, has given a piece of software, such as a mobile app, permission (consent) to act on their behalf. 

> Open Payments requires that consent is collected via an interactive grant before an <LinkOut href="https://openpayments.dev/introduction/op-concepts/#outgoing-payment">outgoing payments</LinkOut> request is issued. A grant is interactive when explicit interaction by an entity (e.g., the software's end user) is required to approve or deny the grant. Tapping an *Approve* button to authorize a payment is an example of an explicit interaction.

Reviewer: ```suggestion
Open Payments requires that consent is collected via an interactive grant before an <LinkOut href="https://openpayments.dev/introduction/op-concepts/#outgoing-payment">outgoing payments</LinkOut> request is issued. A grant is interactive when explicit interaction by a resource owner (e.g., the software's end user) is required to approve or deny the grant. Tapping an *Approve* button to authorize a payment is an example of an explicit interaction.
```

> Open Payments requires that consent is collected via an interactive grant before an <LinkOut href="https://openpayments.dev/introduction/op-concepts/#outgoing-payment">outgoing payments</LinkOut> request is issued. A grant is interactive when explicit interaction by an entity (e.g., the software's end user) is required to approve or deny the grant. Tapping an *Approve* button to authorize a payment is an example of an explicit interaction.

Reviewer: A small, but potentially useful distinction:

> Open Payments requires that consent is collected via an interactive grant before an outgoing payments request is issued

The Open Payments API itself doesn't mandate any grants to be interactive per se (it just says a grant request request can be non-interactive or interactive). The opinionated Open Payments AS in Rafiki _does_, however.

Interactive grants can be optional for incoming payments and quotes; however, they're enabled by default in Rafiki (the `LIST_ALL_ACCESS_INTERACTION` environment variable is `true`). When a grant request includes a `list-all` action for incoming payments and quotes, the request requires interaction. The `list-all` action is used when the client asks to list resources that it did not create.

> If `LIST_ALL_ACCESS_INTERACTION` is `false`, you can still force interactive grants for quotes by setting `QUOTE_INTERACTION` to `true`.

Reviewer: We can add the same for `INCOMING_PAYMENT_INTERACTION`

See the <LinkOut href="https://openpayments.dev/introduction/grants/">Open Payments documentation</LinkOut> for more information on grant negotiation and authorization.

## Authorization servers

Authorization servers grant permission to clients to access the Open Payments Resource APIs. This enables clients to create incoming payments, quotes, and outgoing payments against an account holder's account.

> Rafiki's [auth service](/integration/services/auth-service) provides you with a reference implementation of an Open Payments authorization server. You can use the service as an alternative to developing your own in-house solution. Additionally, the server extends an [API](#interaction-endpoints) that provides interaction endpoints for your IdP.

Reviewer: > You can use the service as an alternative to developing your own in-house solution.

Maybe we can say we _recommend_ you use it?

### Environment variables

The following variables must be configured for the `auth` service.

<div class="overflow-table wide-column">

| Variable | Helm value name | Default | Description |
| -------- | --------------- | ------- | ----------- |

> | `IDENTITY_SERVER_URL` | `auth.identityServer.domain` | N/A | The URL of your IdP's server, used by your authorization server to inform an Open Payments client of where to redirect the end-user to complete the interaction and authorize a grant. | Y | 

Reviewer: ```suggestion
| `IDENTITY_SERVER_URL` | `auth.identityServer.domain` | N/A | The URL of your IdP's server, used by the authorization server to inform an Open Payments client of where to redirect the end-user to complete the interaction and authorize a grant. | Y | 
```

Since we are talking about configuring the Rafiki auth service here

| <Badge text="GET" variant="note" size="medium"/> | `/grant/{id}/{nonce}` | [Look up grant information](#look-up-grant-information) | Identity provider | No |
| <Badge text="POST" variant="success" size="medium"/> | `/grant/{id}/{nonce}/{choice}` | [Accept or reject grant](#accept-or-reject-grant) | Identity provider | No |
| <Badge text="GET" variant="note" size="medium"/> | `/interact/{id}/{nonce}/finish` | [Finish user interaction](#finish-interaction) | Identity provider | Yes |
| <Badge text="POST" variant="success" size="medium"/> | `/interact/{id}/{nonce}` | [Continue grant](#continue-grant) | Open Payments client | Yes |

</div>

We also provide an <LinkOut href="https://github.com/interledger/rafiki/blob/main/packages/auth/src/openapi/specs/id-provider.yaml">OpenAPI specification</LinkOut> that describes the endpoints. Note that the *Continue grant* endpoint is not included in the spec because it's part of the <LinkOut href="https://openpayments.dev/apis/auth-server/operations/post-continue/">Open Payments Auth Server API</LinkOut>.

#### Start user interaction session

> Called by the client to establish an interactive session with your authorization server. Also redirects the client's browser session to your IdP consent screen.

Reviewer: Something like "...with your authorization server, which automatically redirects the request to your IdP consent screen (set by `IDENTITY_SERVER_URL`)".

#### Look up grant information

Called by your IdP server to retrieve a list of access rights, requested by the client, from your authorization server. The request is secured with an [`x-idp-secret`](#x-idp-secret-header) header. The access rights are presented to the client's end-user on the consent screen. The authorization server's response is served on your configured `INTERACTION_PORT`.

#### Accept or reject grant

> Your IdP server communicates the choice made by the end-user on the consent screen (accept/reject) to your authorization server. The request is secured with an [`x-idp-secret`](#x-idp-secret-header) header. Then, your IdP server calls the `finish` endpoint on Rafiki's `auth` service, which handles the interaction requests on your defined `INTERACTION PORT`.

Reviewer: The part about `finish`, I'm not sure whether it belongs here on in the section below

Author: @mkurapov Do you have a preference for the wording here? One is specific to Rafiki and the other is more general to Open Payments:

Called by your IdP server to end the interaction and redirect the end-user's browser session to the...

* Option 1: `finish` endpoint on Rafiki's `auth` service, which handles the interaction requests on your defined `INTERACTION PORT`.
* Option 2: `finish` URI that was provided in the <LinkOut href="https://openpayments.dev/apis/auth-server/operations/post-request/">grant initialization request</LinkOut>.

Reviewer: I think Option 2 is clear:

"Called by your IdP server to end the interaction. If there was a `finish` URI provided in the original grant initialization request, the authorization server redirects the user to that URI".

For the `INTERACTION_PORT`, I think it's sufficient to just mention it once in the "Interaction endpoints" section

Reviewer: Also, for the "Your authorization server automatically redirects the end-user’s browser session to the finish endpoint on Rafiki’s auth service" sentence below, it seems as though "Your authorization server" and "Rafiki’s auth service" are different things perhaps

#### Finish interaction

> Called by your IdP server to end the interaction and redirect the end-user's browser session to the URI of the <LinkOut href="https://openpayments.dev/apis/auth-server/operations/post-request/">grant initialization request</LinkOut>.

Reviewer: Same as the interaction start, we can say the AS automatically redirects the end-user back to the finish URI of the original grant request by the client.

> Called by your IdP server to end the interaction and redirect the end-user's browser session to the URI of the <LinkOut href="https://openpayments.dev/apis/auth-server/operations/post-request/">grant initialization request</LinkOut>.

Reviewer: For the "Your authorization server automatically redirects the end-user’s browser session to the finish endpoint on Rafiki’s auth service" section, it seems as though "Your authorization server" and "Rafiki’s auth service" are different things perhaps

<div class="overflow-table wider-column">

| Response | Description | Example | 
| -------- | ----------- | ------- |
| Rejected | The interaction was rejected by the end-user | `?result=grant_rejected` |
| Invalid  | The grant was not in a state where it could be accepted or rejected (e.g., grant was already approved) | `?result=grant_invalid` |
| Success  | The grant was successful with the following returned in the response:<br /><ul><li>A hash representing the SHA-256 hash of values provided by the client in the <LinkOut href="https://openpayments.dev/apis/auth-server/operations/post-request/">grant initialization request</LinkOut> (`interact.finish.nonce`), and the values in the response returned from your authorization server (`interact.finish`).</li><li>The `interact_ref` that identifies the interaction on your authorization server alongside the hash</li><li>The URI of the grant initialization request (e.g., `https://www.auth-server.com`)</li></ul> | `hash=p28jsq0Y2KK3WS__a42tavNC64ldGTBroywsWxT4md_jZQ1R\HZT8BOWYHcLmObM7XHPAdJzTZMtKBsaraJ64A &interact_ref=4IFWWIKYBC2PQ6U56NL1` |

</div>

> When successful, the SHA-256 hash of the interaction is sent in the response to the client, along with an `interact_ref` that identifies the interaction on the authorization server and the URI of the grant initialization request. The client must verify the hash before it will request to continue the grant

Reviewer: ```suggestion
When successful, the SHA-256 hash of the interaction is sent in the response to the client, along with an `interact_ref` that identifies the interaction on the authorization server and the URI of the grant initialization request. The client must verify the hash before it will request to continue the grant.
```

#### Continue grant

> The client requests a grant from your authorization server for an accepted interaction. Your authorization server responds with an <LinkOut href="https://openpayments.dev/apis/auth-server/operations/post-continue/">access token</LinkOut>.

Reviewer: Technically speaking, there is a single grant (ie the grant was initially requested, but needed a interaction to complete before "continue" was called, and the grant access token be issued to the client).

Author: This section no longer makes mention of grants or access tokens. It now focuses on the header itself and its purpose (securing communications).

- `POST /grant/:id/:nonce/accept`
- `POST /grant/:id/:nonce/reject`

> The header's purpose is to secure communications between the IdP and authorization servers. Its value should be a shared secret known to both entities. When the IdP server sends requests to the authorization server, the IdP must provide the secret via this header.

Reviewer: ```suggestion
The header's purpose is to secure communications between the IdP and the authorization server. Its value should be a shared secret known to both entities. When the IdP server sends requests to the authorization server, the IdP must provide the secret via this header.
```
this one is specific to Rafiki AS
